funcx_sdk/funcx/sdk/asynchronous/ws_polling_task.py

```diff
@@ -134,25 +134,31 @@ async def send_outgoing(self, queue: asyncio.Queue):
             task_group_id = await queue.get()
             await self.ws.send(task_group_id)
 
-    async def handle_incoming(self, pending_futures, auto_close=False) -> bool:
+    async def handle_incoming(self, pending_futures: t.Dict[str, FuncXTask]) -> bool:
         """
-
         Parameters
         ----------
         pending_futures
-        auto_close
 
         Returns
         -------
         True -- If connection is closing from internal shutdown process
         False -- External disconnect - to be handled by reconnect logic
         """
+        # Handle previously-unknown received results (from the .batch_run() case)
+        unprocessed_task_ids = self.unknown_results.keys() & pending_futures.keys()
+        for task_id in unprocessed_task_ids:
+            data = self.unknown_results.pop(task_id)
+            if await self.set_result(task_id, data, pending_futures):
+                return True
+
```

Comment:
This moves this block out of the main loop here. So now it only executes once, rather than "once per tick".

My understanding is that `unknown_results` is populated with data when a result is received before `add_task` is called. e.g. If a task is submitted and added to the poller "right away" but the task executes and returns a result on this path very quickly, before `add_task` gets called.

So aren't we losing something important if we take this out of the loop?

Reply:
The initial observation was that this case is covered because `pending_futures` only gets smaller for the duration of this method execution.  I've just traced `pending_futures` and looks that assumption is false: `pending_futures` can be added to by another thread (via `FuncXExecutor._function_future_map` starting from `FuncXExecutor.start_batching_thread()` )

Thank you for that catch.